### ISSUE TITLE
docker + Nvim-R

### ISSUE THREAD

**Issue by kprimice** [jalvesaq/Nvim-R#259]
I am trying to use Nvim-R with R inside a docker container.
I used this "trick" which I find a big ugly:

```
let R_app = "docker"

let R_args =['run -it --rm', '-e DISPLAY=$DISPLAY', '-v /tmp/.X11-unix:/tmp/.X11-unix:ro', '--ipc host my-r-image']
```

Basic communication between vim and R works, I can send code with the space bar. But that's all I can do : 

- I have to install the old setwidth package so that the width of the console adjusts itself
- The Object Browser doesn't work
- Copy/Pasting chunk of code doesn't work

For the last point I think I just have to share the /tmp file between the docker container and the host. But for everything else I don't know.

Maybe an easier solution would be to put botch Neovim and Nvim-R inside the same container?

**Comment by jalvesaq** [comment]
Please, read section "5.5. R must be started by Vim" of Nvim-R documentation. It looks like that docker is not passing the environment variables that nvimcom needs. From your command above, I infer that `-e` means "copy environment variable". Then, use it to copy `NVIMR_TMPDIR`, `NVIMR_COMPLDIR`, `NVIMR_ID`, and `NVIMR_SECRET`.

**Comment by jalvesaq** [comment]
You also have to copy `R_DEFAULT_PACKAGES`...

**Comment by kprimice** [comment]
Thanks, mounting `NVIMR_TMPDIR` did the job for the copy/paste. I guess mounting `NVIMR_COMPLDIR` is required too.

Also nvimcom is required inside the docker container. But because NVim-R is installed on the host, and takes care of installing nvimcom, nvimcom is installed on the host and not inside the container.
Is there a way to install nvimcom manually inside the container ? (A quick fix would be to copy the repository, but I'd like the container to be as independent to the host as possible).

**Comment by jalvesaq** [comment]
I don't know exactly what docker is. My suggestions here and in vimcmdline were entirely based on your own description of the problem...
If you mount the two directories (TMPDIR and COMPLDIR), and if you pass to R all five environment variables, nvimcom should be automatically installed just before starting R whenever necessary. You can also have both the directory where temporary files are created and the directory where completion data is cached in the remote machine, but this requires additional configuration. The discussion at https://github.com/jalvesaq/Nvim-R/issues/184 might be relevant to you. Please, read:
```
:h Nvim-R-remote
```

**Comment by kprimice** [comment]
Thanks. Looking at your Diagram, I think only the black path works for me.

- nclientserver is not running (,rh doesn't work for example)
- When opening an helper file in the R console, it doesn't open in an R buffer
- The GlobalView is not working

I guess it's the TCP connection that fails. I'm not sure I can use the `NVIM_IP_ADDRESS` in my case.

Also TMPDIR and COMPLDIR are mounted and have exactly the same path on botch machines.

vimcom doesn't seem to be installed automatically on the remote machine (Here the path to the R_library is different)

**Comment by kprimice** [comment]
After a reboot, when I try to launch R with docker from neovim I get an error:

> The input device is not a TTY

It was working properly before the reboot and:

- It works when R is launched normally (not inside a docker container)
- When not launched from vim, R works inside a docker container

Also the TERM env variable is set.

**Comment by jalvesaq** [comment]
I have found this: https://stackoverflow.com/questions/43099116/the-input-device-is-not-a-tty

**Comment by kprimice** [comment]
Yes, but I need both `i` and `t` with Nvim-R. (I need an interactive terminal).

I tried to use your other plugin `vimcmdline` with R and I managed to launch a Docker container with R without error.
vimcmdline also works remotely with javascript: Launching a docker container on a remote server.

My clue is that it comes from the command that execute `R_cmd`. I found this: https://stackoverflow.com/questions/44577344/piping-docker-run-container-id-to-docker-exec
You get the same error with xargs because the tty is not passed.

**Comment by jalvesaq** [comment]
I don't know how to fix the issue, but, please, report here if you find a solution.

**Comment by kprimice** [comment]
I managed to make it work. But it's dirty: I removed `CheckNvimcomVersion()` from `common_global.vim`.

The line 559 was the problem. Since I'm running inside tmux (not in a buffer), I thought it would reach line 839, but it never did.

So it works as before (see messages above) because I never managed to fully handle the communication between vim and the container.
The scripts waits for nvimcom, never finds it, prints an error message, and then I'm good to go.

But anyway it was weird that I got this "The input device is not a TTY" error out of the blue.

**Comment by jalvesaq** [comment]
If you are using Vim, could you try Neovim? I noted on Windows that Vim pops up a lot of CMD windows while starting up with Nvim-R enabled while Neovim does not. It seems that Vim and Neovim execute the `system()` command in different ways. Perhaps, Neovim does not need a TTY to run `system()`.

**Comment by kprimice** [comment]
Sorry I should have mention that:

- I'm using Neovim
- I'm running linux (Debian)

**Comment by jalvesaq** [comment]
That call to `system()` was the only one with the `input` argument. I've just changed the R command to avoid that. Could you try again, please?

**Comment by kprimice** [comment]
I still get the same error

**Comment by jalvesaq** [comment]
Unfortunately, I don't know what else I could do...

**Comment by kprimice** [comment]
No problem, I will look more into it next week when I find the time!

**Comment by kprimice** [comment]
I made some progress. First of all, I noticed that R_IP_ADDRESS and NVIM_IP_ADDRESS were wrong inside the Docker container. So I made some changes in my .Rprofile:

> 
> 	if(Sys.getenv("NVIM_IP_ADDRESS") != ""){
> 			Sys.setenv("R_IP_ADDRESS" = trimws(system("awk 'END{print $1}' /etc/hosts", intern = TRUE)))
> 			Sys.setenv("NVIM_IP_ADDRESS" = trimws(system("/sbin/ip route|awk '/default/ { print $3 }'", intern = TRUE))) 
> 	}

Which sets both IPs correctly.

But I still can't get nvimcom/nclientserver to work correctly. (I don't have completion, and the Rdoc doesn't open in nvim even if it is actually created inside the container).

R_tmpdir seems to work, at it is exactly the same on both ends.

Here are my questions:
R_compldir is more complicated. It is different on both machines by default (Because the username is different, so the path to the home is different too).
Also, if I try to set R_compldir in my vimrc, R doesn't launch anymore. Without any error message. Any idea why?

**Comment by jalvesaq** [comment]
You can put in your `.Rprofile` the command `options(nvimcom.verbose=4)` to see whether it is receiving Nvim-R messages or not. If R receives the messages, then, it will write the files in the temporary directory. If the temporary directory is really the same for both R and Nvim-R, then Vim should successfully read the files. Is `/tmp` for Vim the same `/tmp` in the container, or does the container create its own file system structure?

**Comment by kprimice** [comment]
The /tmp directory works fine actually (copy/pasting from nvim to the console works for example). The /tmp from the host is mounted on the container so they share the exact same file and I didn't have to change R_tmpdir since the absolute path is the same on the host and on the container.

I can't replicate the same thing with R_compldir because it is in the /home directory. And the user name of the host and the container may differ. The easiest way should be to change R_compldir so that is it also in /tmp (for example). But when I try to set R_compldir in my .vimrc R won't launch anymore.
Same thing happen if set R_compldir to the default path.

Also, what ports nvimcom uses? Is it 10104?

The problem seems to be with R_compl

**Comment by jalvesaq** [comment]
`nvimcom` uses the first port available from 10000 to 10048, and `nclientserver` uses the first one from 10100 to 10148.

**Comment by kprimice** [comment]
Until now I was using Neovim and tmux. But because of your last commit I thought I would have better chances to succeed by switching to Neovim only.

However, after some tweaking, I now get this error:

> Error: package or namespace load failed for ‘nvimcom’:                                                                                                                                                                                         
 .onLoad failed in loadNamespace() for 'nvimcom', details:                                                                                                                                                                                     
  call: setwd("/home/localuserhome")                                                                                                                                                                                                           
  error: cannot change working directory

So the home path is wrong and should be "/home/remoteuserhome"
But I can't find where it is defined or how to change it

**Comment by jalvesaq** [comment]
Please, try different values for `R_nvim_wd`.
See `:h R_nvim_wd`.

**Comment by kprimice** [comment]
I set it to -1

Now I get this at startup (with verbose for nvimcom)

> nclientserver port:                                                                                                                                                                                                                            
nvimcom 0.9.64 loaded                                                                                                                                                                                                                          
    NVIMR_TMPDIR = /tmp/Nvim-R                                                                                                                                                                                                           
    NVIMR_ID = 116998040831400601                                                                                                                                                                                                              
R_IP_ADDRESS: 172.17.0.46                                                                                                                                                                                                                      
nvimcom port: 10001                                                                                                                                                                                                                            
nvimcom_nvimclient(SetNvimcomInfo('0.9.64', '/usr/local/lib/R/site-library/nvimcom', '10001', '1', '0', 'nvimcom stats graphics grDevices utils datasets ggplot2 colorout setwidth methods base Dec.', '3.4.3')): '' (0)                       
nvimcom_nvimclient() called although Neovim server port is undefined 

Apparently there is no port defined. It seems to be related to NVIMR_PORT but I can't find anything in the doc.

**Comment by jalvesaq** [comment]
Some time ago I could successfully run Neovim on a local machine and R in a remote machine. The required configuration that worked at the time is described in the section "Remote access" of Nvim-R documentation. I can't guarantee that it would still work, and I don't know if running R in a docker session is the same thing, but, perhaps, reading the section might be useful. If you find a way of making Nvim-R work within a docker session, please, post the solution here.

**Comment by kprimice** [comment]
I've read that section a couple of times.

> You have to edit your `~/.Rprofile` to create the environment variable
	`R_IP_ADDRESS`. R will save this value in a file that Vim has to read
	to be able to send messages to R.

Where is that file? (Where R_IP_ADDRESS is stored).
Or where should I write the IP address?

That IP is generated at the same time as the R session. So I can't know it beforehand. But I can paste it while the session is created.

**Comment by jalvesaq** [comment]
The command is in the documentation:
```
if(interactive() && Sys.getenv("NVIM_IP_ADDRESS") != ""){                                                                                                             
    Sys.setenv("R_IP_ADDRESS" = trimws(system("hostname -I", intern = TRUE)))                                                                                         
}
```
It's not a file. It's a environment variable set by `Sys.setenv()`.

**Comment by kprimice** [comment]
Yes I saw this command. However there is still something I don't understand. The "R_IP_ADDRESS" env variable is used by the localhost (vim) to communicate with R, right?
So it should be set on the host, not on the remote?

The command in the doc sets "R_IP_ADDRESS" on the remote, not the host. (R_IP_ADDRESS and NVIM_IP_ADDRESS are correctly defined on my remote/container, however, as seen above, nclientserver port is undefined. I guess it is undefined because it can't scan for the port, because it doesn't have the IP of the remote)

**Comment by jalvesaq** [comment]
This is a configuration that I never need, and there were some changes in the nclientserver initialization in the last weeks. I'm sorry, but I would have to spend hours checking if the "Remote" section of Nvim-R documentation is still valid, and, unfortunately, I can't look at this now.

**Comment by kprimice** [comment]
Could you describe to me how the R/Nvim communication is supposed to work in the remote configuration?

- R loads nvimcom that starts a TCP client and server (remote)
- nclientserver starts a TCP client and server (host)
- nvimcom sends the SetNvimcomInfo through TCP to Nvim. Nvim executes it, and then the communication is fully operational.

Is that correct?

**Comment by jalvesaq** [comment]
The sequence of R startup on a local machine  is:

1. Check if nvimcom is installed and if it is up to date. Install/update nvimcom if necessary.
2. Start the `nclientserver` application.
3. Start R which will load nvimcom.

The problems for a local Vim/Neovim and a remote R are:

- Step 1 will work only if nvimcom is already installed and up to date on both the local machine and the remote machine.
- Step 2 will work only if the `nclientserver` application can be found on the local machine. That is why nvimcom must be compiled on the local machine too.
- On step 3, R will load nvimcom only if the environment variable `R_DEFAULT_PACKAGES` exists and includes ",nvimcom", and if both the local and the remote machines share the same temporary Nvim-R directory.

However, R (and, consequently, nvimcom) used to be started before nclientserver. Now, it is the opposite.

So, it will be really easier if you could run Neovim in a docker session and, then, start R there with `<LocalLeader>rf`.

**Comment by kprimice** [comment]
I have finally a bit of time to come back to this. (I've been using a dirty fix for the past few months but I'd like to make a clean pull request).

I have a few questions:

- Omnicompletion seems to be an issue, I get this error at launch: 

```
nvimcom:::nvim.buildomnils("datasets")
Error in if (getOption("nvimcom.verbose") > 3) cat("nvimcom R: omnils does not exist:",  : 
  argument is of length zero
Calls: <Anonymous>
Execution halted
```
Code: https://github.com/jalvesaq/Nvim-R/blob/master/R/common_global.vim#L620
"nvimcom.verbose" is not defined (logcal(0)) in my case.

- Bypassing previous error it then can't find nvimcom/bin : https://github.com/jalvesaq/Nvim-R/blob/master/R/common_global.vim#L669
Does this path should be the one on the host, or on the remote machine ?
Because `rplugin_compldir` is set to the remote path, and is used here as a local path. Which is why it can't find the bin file.

**Comment by jalvesaq** [comment]
The option nvimcom.verbose is defined when nvimcom is being loaded: https://github.com/jalvesaq/Nvim-R/blob/master/R/nvimcom/R/nvimcom.R#L18

**Comment by jalvesaq** [comment]
I have options(nvimcom.verbose = 1) in my .Rprofile...

**Comment by jalvesaq** [comment]
The instructions at section 9.21 of Nvim-R documentation include `R_DEFAULT_PACKAGES` among the environment variables that have to be exported to run R remotely.

**Comment by jalvesaq** [comment]
I changed the code to ensure that nvimcom.verbose is set in nvim.buildomnils, but you have to delete ~/.cache/Nvim-R/nvimcom_info to force the update of nvimcom because I didn't increase its version number.

**Comment by kprimice** [comment]
Now I get this error:

```
  > nvimcom:::nvim.buildomnils("datasets")
  Error in file(con, "w") : cannot open the connection
  Calls: <Anonymous> -> nvim.bol -> writeLines -> file
  In addition: Warning message:
  In file(con, "w") :
    cannot open file '/pack_descriptions': Permission denied
  Execution halted
```

`R_DEFAULT_PACKAGES` is exported and `buildomnils.R` is created with the following content:

```
nvimcom:::nvim.buildomnils("datasets")
nvimcom:::nvim.buildomnils("utils")
nvimcom:::nvim.buildomnils("grDevices")
nvimcom:::nvim.buildomnils("graphics")
nvimcom:::nvim.buildomnils("stats")
nvimcom:::nvim.buildomnils("methods")
nvimcom:::nvim.buildomnils("nvimcom")
nvimcom:::nvim.buildomnils("base")
```

**Comment by kprimice** [comment]
Nevermind, I was passing `NVIM_COMPLDIR` only if `NVIM_ID` was set.

So this part (omni completion) is fine now. And I get again:

> Application "nclientserver" not found.

**Comment by kprimice** [comment]
I changed the code to ensure that if R_nvimcom_home is set then it should be checked first and used if it exists.

**Comment by kprimice** [comment]
So now R launches properly in an nvim buffer. However the communication between nvim and R is broken. I get:

> Not ready yet

when I try so send a line of code to R.

And after a while:

> The package nvimcom wasn't loaded yet. Please, quit R and try again.

Where do you think I should look?

**Comment by kprimice** [comment]
From `Sys.getenv()`

> R_DEFAULT_PACKAGES      datasets,utils,grDevices,graphics,stats,methods,nvimcom

So it should be fine, right?

**Comment by jalvesaq** [comment]
> So it should be fine, right?

Yes.

Then, the problem is that messages sent by nvimcom are not reaching Vim/Neovim.

**Comment by jalvesaq** [comment]
Did you set `NVIM_IP_ADDRESS` too?

**Comment by kprimice** [comment]
Yes it is also set (and `R_IP_ADDRESS` too)

Could you point out where I should look in the code? I didn't find the relevant function(s). So that I can try to debug it myself

**Comment by jalvesaq** [comment]
You can put in your .Rprofile:

```
options(nvimcom.verbose = 4)
```

You can also set the environment variable `DEBUG_NVIMR` and use `tail` to follow the files `/tmp/nclientserver_debug` and `/tmp/nclientserver_1_debug`:

```sh
tail -f /tmp/nclientserver_debug
```

and

```sh
tail -f /tmp/nclientserver_1_debug
```

The communication between R and Vim is intermediated by nvimcom and nclientserver:

```
R/nvimcom/src/nvimcom.c
R/nvimcom/src/apps/nclientserver.c
```
You could, then, compare R running locally and remotely.

**Comment by jalvesaq** [comment]
I don't know much about sockets, but, perhaps, the file created after `?ls` although present in the remote machine wasn't flushed through the network. You must be sure that Vim is looking for the temporary file at the correct place.

I can't run a docker setup, and I'm sorry but I have no other idea of what you could do to establish the communication between Vim and R.

**Comment by kprimice** [comment]
I think the problem comes from `R_IP_ADDRESS` that is not properly set. (I thought it was but it is only set in the docker container, which is useless, right?).

If I understand correctly, it must be set in the nvim environment.

However I can only know the IP of the container *after* R is launched. (with a simple shell command).

**Comment by kprimice** [comment]
Found the solution. I just have to set `R_IP_ADDRESS` to a non-NULL value in my vimrc.

Everything seems to work perfectly now! Thank you for your help

**Comment by kprimice** [comment]
I am reopening this issue because it does not properly work with 2 NVim and 2 R sessions open. Here is what happens:

1. I launch Session A from NVim A. Everything works as expected and communication between Vim and R works perfectly.
2. I launch Session B from NVIM B.
3. NVim A displays an error 

> [ClientServer] Strange string received: "1733109729674567458call SetNvimcomInfo('0.9.75'.....

4. I can't send a command to R Session B. It says "Not ready yet".
5. If I write "?ls" in R Session B I also get a ClientServer error in NVIM A.

Some debug infos about R Session A:

> >   nclientserver port: 10101
> nvimcom 0.9.75 loaded
>     NVIMR_TMPDIR = /home/docker/.cache/Nvim-R/tmp
>     NVIMR_ID = 1973899965102194721
> R_IP_ADDRESS: 172.17.0.3
> nvimcom port: 10001

And R Session B:

> nclientserver port: 10101
> nvimcom 0.9.75 loaded
>     NVIMR_TMPDIR = /home/docker/.cache/Nvim-R/tmp
>     NVIMR_ID = 1315723886533425404
> R_IP_ADDRESS: 172.17.0.6
> nvimcom port: 10001

Are nclientserver and nvimcom supposed to run on the same ports for both sessions?

**Comment by jalvesaq** [comment]
> Are nclientserver and nvimcom supposed to run on the same ports for both sessions?

No. If the ports are already in use, they try the next port.

**Comment by ErickChacon** [comment]
@kprimice, can you make a summary  of how to use docker + Nvim-R?
